### DISCUSSION TITLE
Loop over all components of all entities from view and call a function by name

### DISCUSSION THREAD

**Discussion by Cyndeon** [skypjack/entt#1175] [Q&A] [answered]
I hope it's alright to ask this here since I could not find the answer anywhere else.

Currently I am working on a tiny custom C++ engine and what I want to do is make a hierarchy where I can click on an object in order to modify its variables.
The way I want to do this, is to loop through every single component saved in the meta registry to check if any of the components from the entities received from view is one that can be modified and thus should be shown.
All of my components that can be modified (which currently are called Emitter and Transform) have a function called DrawInspector().
So when a component that is saved in the meta registry has been recognized on the entity that has this function, I want to call it.
Below I will show what code I have and what I still need.

Here I set up the meta data
```
// set up the Registry's meta data
entt::meta<bee::components::Transform>()
    .type(entt::hashed_string("Transform"))
    .func<&bee::components::Transform::DrawInspector>(entt::hashed_string("DrawInspector"));
entt::meta<bee::custom::Emitter>()
    .type(entt::hashed_string("Emitter"))
    .func<&bee::custom::Emitter::DrawInspector>(entt::hashed_string("DrawInspector"));
```

This is my drawUI function. As you can see, I have a seperate window for this and I check the view for the empty struct called SerializableTraits. I then go through the entire list of meta data types, then construct an empty class and call the function.
It is an empty class since at least this will build, but what I need is to call that on the components themselves, not on the types, but I cannot figure out how to do that at all. I've been trying for hours and asked help from others to no avail.

```
ImGui::Begin("Inspector", nullptr, 0);

ImGui::SetWindowSize(ImVec2(400.0f, 600.0f));
{
    auto view = Engine.Registry().view<SerializableTraits>();
    for (auto entity : view)
    {
        std::string header = "Entity" + std::to_string(static_cast<int>(entity));
        ImGui::PushID(static_cast<int>(entity));
        if (ImGui::CollapsingHeader(header.c_str()))
        {
            for (auto&& [id, type] : entt::resolve())
            {
                // way to get name, but is unused
                // std::string fullName(type.info().name());
                //// make sure to filter out any namespaces
                // if (fullName.find(':'))
                //{
                //     auto lastPos = fullName.find_last_of("::") + 1;
                //     std::string fillerName = "";
                //     for (size_t i = lastPos; i < fullName.length(); i++)
                //     {
                //         fillerName += fullName[i];
                //     }
                //     fullName = fillerName;
                // }

                if (auto func = type.func(entt::hashed_string("DrawInspector")); func)
                {
                    // this has to be replaced with something like view.get<type>(entity) but this one doesn't work
                    auto any = type.construct();
                    if (any)
                    {
                        func.invoke(any);
                    }
                }
            }
            // stuff that does work but is not as clean nor as easy to expand
            //{
            ///////////if (Emitter* type = Engine.Registry().try_get<Emitter>(entity); type != nullptr)
            /// type->DrawInspector();
            ///////////if (Transform* type = Engine.Registry().try_get<Transform>(entity); type != nullptr)
            /// type->DrawInspector();
        }
        ImGui::PopID();
    }
}
ImGui::End();
```

I really hope someone can help me out or tell me that what I am trying to do is impossible :D
I also feel like I should mention, no virtual voids are allowed for this since it is an engine and thus has to be performant.
I really hope this is not a duplicate since I did plenty of research but please let me know if there is a method that worked before and actually still works (since lots of links are outdated and don't work anymore).

**Comment by skypjack** [comment, marked as answer]
> I also feel like I should mention, no virtual voids are allowed for this since it is an engine and thus has to be performant.

There is a `value(entity)` function on the storage, you can combine it with `type.from_void(...)` to get a `meta_any` on which to invoke `func`.
However, I get it doesn't work for you due to a hypothetical performance problem that was never measured but only imagined? 🤔 
In this case, the other alternative is to attach yet another function to the meta type to get the component from the registry.
My two cents: I'd use the cooked solution first, measure and only change it if you spot a bottleneck there.

**Comment by Cyndeon** [comment]
Thank you for the answer. I actually really wanted this to be expandable in the future, since it is supposed to be an engine and users likely would not have access to the library itself, thus, if this was hard-coded, users could not add their own types unless I'd use something like delegates, which, while possible, I feel like would clutter the code more when combined with all the hard-coded lines for all existing components.
I will however look into the storage you mentioned and attempt to use that :D

**Reply by skypjack** [reply]
Feel free to reach out if you have problems. It's not that difficult though, something along this line at the end of the day:
```cpp
for (auto [id, storage] : registry.storage()) {
    auto type = entt::resolve(storage.info());

    if (auto func = type.func(entt::hashed_string("DrawInspector")); func) {
        func.invoke(type.from_void(storage.value(entity)));
    }
}
```
I'm pretty sure it doesn't work out of the box 🙂 but it should give you a grasp of what I mean. 👍
